src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs

```diff
@@ -620,15 +621,31 @@ protected internal override Task<HttpResponseMessage> SendAsync(HttpRequestMessa
                 return Task.FromCanceled<HttpResponseMessage>(cancellationToken);
             }
 
-            HttpMessageHandlerStage handler = _handler ?? SetupHandlerChain();
-
             Exception? error = ValidateAndNormalizeRequest(request);
             if (error != null)
             {
                 return Task.FromException<HttpResponseMessage>(error);
             }
 
-            return handler.SendAsync(request, cancellationToken);
+            // SetupHandlerChain may block for a few seconds in some environments.
+            // (See https://github.com/dotnet/runtime/issues/115301. )
+            // The setup procedure is being enqueued to thread pool to prevent the caller from blocking
+            async Task<HttpResponseMessage> CreateHandlerAndSendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+            {
+                if (Volatile.Read(ref _handlerChainSetupTask) is null)
+                {
+                    _handlerChainSetupTask = Task.Run(SetupHandlerChain);
+                }
+
+                HttpMessageHandlerStage handler = await _handlerChainSetupTask!.ConfigureAwait(false);
+
+                return await handler.SendAsync(request, cancellationToken)
+                    .ConfigureAwait(false);
```

Comment:
```suggestion
                _handlerChainSetupTask ??= Task.Run(SetupHandlerChain);
                HttpMessageHandlerStage handler = await _handlerChainSetupTask.ConfigureAwait(false);
                return await handler.SendAsync(request, cancellationToken).ConfigureAwait(false);
```

Reply:
Could you elaborate why the variant without volatile read is preferrable?
It's more simple, but to my understanding the caller is still before the first yeild, so there are high chances `null` would be seen even if the other thread managed to populate the field.

Reply:
If we wanted to ensure only one init is running here, we'd have to lock. Volatile does not ensure that other threads immediately see the new task, and read reordering isn't a concern within this method.

The race condition isn't problematic here, and is small either way.

Reply:
Thank you. I didn't imply it would be race-free, only less probable, given the side effects of `Volatile.Read` on a single processor systems as I understand them.

I'll apply the suggestion then.